File: rust/test/shared/utf8/utf8_test.cc
#include <string>

#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include "absl/strings/string_view.h"
#include "rust/test/shared/utf8/feature_verify.pb.h"
#include "rust/test/shared/utf8/no_features_proto2.pb.h"
#include "rust/test/shared/utf8/no_features_proto3.pb.h"

namespace {

using ::testing::Eq;
using ::testing::IsEmpty;
using ::testing::Not;

// We use 0b1000_0000, since 0b1XXX_XXXX in UTF-8 denotes a byte 2-4, but never
// the first byte.
constexpr char kInvalidUtf8[] = "\x80";

TEST(Utf8Test, TestProto2) {
  utf8::NoFeaturesProto2 no_features_proto2;

  // No error on setter.
  no_features_proto2.set_my_field(kInvalidUtf8);
  EXPECT_THAT(no_features_proto2.my_field(),
              Eq(absl::string_view(kInvalidUtf8)));

  // No error on serialization.
  std::string serialized_nonutf8 = no_features_proto2.SerializeAsString();
  EXPECT_THAT(serialized_nonutf8, Not(IsEmpty()));

  // No error on parsing.
  utf8::NoFeaturesProto2 parsed;
  EXPECT_THAT(parsed.ParseFromString(serialized_nonutf8), Eq(true));
}

TEST(Utf8Test, TestProto3) {
  utf8::NoFeaturesProto3 no_features_proto3;

  // No error on setter.
  no_features_proto3.set_my_field(kInvalidUtf8);
  EXPECT_THAT(no_features_proto3.my_field(),
              Eq(absl::string_view(kInvalidUtf8)));

  // No error on serialization.
  std::string serialized_nonutf8 = no_features_proto3.SerializeAsString();
  EXPECT_THAT(serialized_nonutf8, Not(IsEmpty()));

  // Error on parsing.
  utf8::NoFeaturesProto3 parsed;
  EXPECT_THAT(parsed.ParseFromString(serialized_nonutf8), Eq(false));
}

TEST(Utf8Test, TestEditionsVerify) {
  utf8::Verify verify;

  // No error on setter.
  verify.set_my_field(kInvalidUtf8);
  EXPECT_THAT(verify.my_field(), Eq(absl::string_view(kInvalidUtf8)));

  // No error on serialization.
  std::string serialized_nonutf8 = verify.SerializeAsString();
  EXPECT_THAT(serialized_nonutf8, Not(IsEmpty()));

  // Error on parsing.
  utf8::Verify parsed;
  EXPECT_THAT(parsed.ParseFromString(serialized_nonutf8), Eq(false));
}

}  // namespace

File: rust/test/shared/utf8/utf8_test.rs
// Protocol Buffers - Google's data interchange format
// Copyright 2024 Google LLC.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

//! Tests demonstrating the Protobuf Rust behavior around UTF-8 under different
//! scenarios (proto2, proto3, editions).

// TODO: The behavior is currently subptimal (for example because
// b/333545903 or b/335140403). Design and implement desirable changes to this
// behavior. Do not assume that the Protobuf team is intentional about these
// behaviors while b/304774814 is open.

use googletest::prelude::*;
use protobuf::prelude::*;

use feature_verify_rust_proto::Verify;
use no_features_proto2_rust_proto::NoFeaturesProto2;
use no_features_proto3_rust_proto::NoFeaturesProto3;
use protobuf::{ParseError, ProtoStr};

// We use 0b1000_0000, since 0b1XXX_XXXX in UTF-8 denotes a byte 2-4, but never
// the first byte.
const NON_UTF8_BYTES: &[u8] = b"\x80";

// Returns ProtoStr with non-UTF-8 content.
fn make_non_utf8_proto_str() -> &'static ProtoStr {
    unsafe {
        // SAFETY: This is safe under current implementation of C++ and UPB kernels.
        // In the hypothethical pure Rust runtime this would be library-level UB - but
        // this test is specifically present to demonstrate UTF-8 behavior under
        // C++ and UPB kernels.
        ProtoStr::from_utf8_unchecked(NON_UTF8_BYTES)
    }
}

#[gtest]
fn test_proto2() {
    let non_utf8_str = make_non_utf8_proto_str();

    let mut msg = NoFeaturesProto2::new();

    // No error on setter
    msg.set_my_field(non_utf8_str);
    assert_that!(msg.my_field().as_bytes(), eq(NON_UTF8_BYTES));

    // No error on serialization
    let serialized_nonutf8 = msg.serialize().expect("serialization should not fail");

    // No error on parsing.
    let parsed_result = NoFeaturesProto2::parse(&serialized_nonutf8);
    assert_that!(parsed_result, ok(anything()));
}

#[gtest]
fn test_proto3() {
    let non_utf8_str = make_non_utf8_proto_str();

    let mut msg = NoFeaturesProto3::new();

    // No error on setter
    msg.set_my_field(non_utf8_str);
    assert_that!(msg.my_field().as_bytes(), eq(NON_UTF8_BYTES));

    // No error on serialization
    let serialized_nonutf8 = msg.serialize().expect("serialization should not fail");

    // Error on parsing.
    let parsed_result = NoFeaturesProto3::parse(&serialized_nonutf8);
    assert_that!(parsed_result, err(matches_pattern!(&ParseError)));
}

#[gtest]
fn test_verify() {
    let non_utf8_str = make_non_utf8_proto_str();

    let mut msg = Verify::new();

    // No error on setter
    msg.set_my_field(non_utf8_str);
    assert_that!(msg.my_field().as_bytes(), eq(NON_UTF8_BYTES));

    // No error on serialization
    let serialized_nonutf8 = msg.serialize().expect("serialization should not fail");

    // Error on parsing.
    let parsed_result = Verify::parse(&serialized_nonutf8);
    assert_that!(parsed_result, err(matches_pattern!(&ParseError)));
}
